### ISSUE TITLE
Cross compiling `goos = darwin` introduces new darwin compiler errors

### ISSUE THREAD

**Issue by achew22** [bazelbuild/rules_go#1148]
In https://github.com/bazelbuild/bazel-watcher/pull/104, I switched over from using `--experimental_platforms=@io_bazel_rules_go//go/toolchain:${GOOS}_${GOARCH}` to using the new `goos` argument in a `go_binary` rule. Unfortunately I now get the following compiler errors.

```
Use --sandbox_debug to see verbose messages from the sandbox
external/org_golang_x_sys/unix/flock.go:16:42: undefined: Flock_t
external/org_golang_x_sys/unix/sockcmsg_unix.go:37:18: undefined: Cmsghdr
external/org_golang_x_sys/unix/sockcmsg_unix.go:43:9: undefined: Cmsghdr
```

I updated `golang.org/x/sys` to the latest version (along with rules_go) in that PR so I don't think it is from version skew. Do you have any hints you could give me on what I might be doing wrong?

**Comment by ianthehat** [comment]
Not sure what is going on there, it looks like you are triggering a cgo build when attempting to cross compile, which is not currently allowed (hopefully soon).
I don't see how you could be though, as you set pure = "on".
What was the host platform?
I will try to replicate the failure.

**Comment by achew22** [comment]
> What was the host platform?

It happens on the Bazel darwin host, ubuntu 16.04, ubuntu 14.04, and my laptop running ubuntu 16.04. Additionally it fails on Travis's "osx" (xcode8) and "linux" images.

**Comment by ianthehat** [comment]
I replicated this in https://github.com/ianthehat/rules_go/tree/cross which I will submit as soon as I find a way to fix it.
I am fairly sure the problem is that the select statements are broken because the toolchain does not change. We are going to have to think about how to fix this, it's not going to be easy.
We might need to rethink whether we even need gazelle to emit select statements for source files. Even if we go that route, we still need to think about deps.
@jayconrod what do you think Jay?

**Comment by jayconrod** [comment]
We discussed this a bit offline.

It doesn't seem like there is a good solution. I'm generally uncomfortable with the `goos` and `goarch` attributes. They're needed because we don't have toolchain transitions (i.e., the ability to depend on a target built for a specific platform). It seems like these will be added to Bazel eventually, but not any time soon. Actually, the native Android rules already have this, but it's just not exposed to the rest of the rules.

Given that we should have a cleaner solution for platform-specific dependencies at some point, I think we should avoid entrenching machine-generated work-arounds. Some solutions we considered were to have Gazelle generate separate `go_source` or `go_library` rules per-platform or have platform specific `srcs` and `deps` attributes. This would quickly produce large build files that are difficult to maintain and difficult to migrate away from.

There are some things we can do to make manual work-arounds easier though. As Ian mentioned above, we don't strictly need `select` expressions in `srcs`, since rules\_go filters source files before compiling them. Filtering these with `select` is slightly more efficient (fewer files to sandbox) but I could add a flag to opt out of that. We definitely need to have `select` expressions for `deps` though; otherwise Bazel will attempt to build (but not link) libraries on platforms that they will never build on.

Another problem is that `go_repository` is getting in the way: the files it generates can't be hand-modified. I'm already planning a replacement for this, but we haven't finalized the design yet.

**Comment by junghoahnsc** [comment]
Is this issue the same as #1174? How I can resolve this issue? 
I'm stick with '2b3efcbfabc464139e37630deef06463aae5f00d' since no other workaround works for me.

**Comment by jayconrod** [comment]
@junghoahnsc It seems like the same issue as #1174. The problem is that when --experimental_platforms is not specified, Bazel will evaluate `select` expressions according to your host platform. So if you're generating cross-platform build files with Gazelle (or `go_repository`) that include `select` expressions, you may get the wrong `srcs` and `deps`. 

From rules_go, we can't really influence how `select` expressions are evaluated, so there's not much we can do to fix this. As a workaround, you would need to write rules that don't include `select` expressions that depend on platforms. If you need the same library to be compiled with different sources on different platforms in the same build, I think you would need to define a `go_library` rule for each variation. Unfortunately, it's a lot of work.

The long term solution to this would be extending the platform / toolchain mechanism in Bazel to support multiple target platforms in the same build. There is some interest in this on the Bazel team (and actually this is already implemented for fat Android NDK binaries), but there are scalability concerns, too.

**Comment by jayconrod** [comment]
Closing some old issues.

The immediate cause of this was a `select` expression that was picking the wrong sources. The mechanism we use for the `goos` and `goarch` attributes (an aspect which re-analyzes everything) does not work with `select` statements because it propagates across the graph after `select` has been evaluated. We mitigated this by making Gazelle generate flat `srcs` list. Sources are now filtered at execution time instead of load time. However, `deps` lists may still contain `select` expressions, and that can result in similar errors. I don't want to flatten `deps` lists because that introduces a new kind of bug: we may fetch additional repos and attempt to build targets that are not actually buildable on our target platform.

There is one more related problem, #1592. cgo is disabled by default for cross-compiling modes, but it may be re-enabled with `pure = "off"`. Our aspect doesn't propagate through cgo rules, so cgo code may be compiled for the wrong target architecture. #1595 would fix this, but it makes analysis ~15% slower, so I want to investigate other solutions before merging it.

`--platforms=@io_bazel_rules_go//go/toolchain:<goos>_<goarch>` is the best way to cross-compile at the moment.
